coral-common/src/main/java/com/linkedin/coral/common/transformers/SqlCallTransformer.java

```diff
@@ -0,0 +1,95 @@
+/**
+ * Copyright 2017-2023 LinkedIn Corporation. All rights reserved.
+ * Licensed under the BSD-2 Clause license.
+ * See LICENSE in the project root for license information.
+ */
+package com.linkedin.coral.common.transformers;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.sql.SqlCall;
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.SqlNodeList;
+import org.apache.calcite.sql.SqlSelect;
+import org.apache.calcite.sql.validate.SqlValidator;
+
+
+/**
+ * Abstract class for generic transformations on SqlCalls
+ */
+public abstract class SqlCallTransformer {
+  private SqlValidator sqlValidator;
+  private final List<SqlSelect> topSelectNodes = new ArrayList<>();
+
+  public SqlCallTransformer() {
+
+  }
+
+  public SqlCallTransformer(SqlValidator sqlValidator) {
+    this.sqlValidator = sqlValidator;
+  }
+
+  /**
+   * Predicate of the transformer, it’s used to determine if the SqlCall should be transformed or not
+   */
+  protected abstract boolean predicate(SqlCall sqlCall);
+
+  /**
+   * Implementation of the transformation, returns the transformed SqlCall
+   */
+  protected abstract SqlCall transform(SqlCall sqlCall);
+
+  /**
+   * Public entry of the transformer, it returns the result of transformed SqlCall if `condition(SqlCall)` returns true,
+   * otherwise returns the input SqlCall without any transformation
+   */
+  public SqlCall apply(SqlCall sqlCall) {
+    if (sqlCall instanceof SqlSelect) {
+      this.topSelectNodes.add((SqlSelect) sqlCall);
+    }
+    if (predicate(sqlCall)) {
+      return transform(sqlCall);
+    } else {
+      return sqlCall;
+    }
+  }
+
+  /**
+   * To get the RelDatatype of a SqlNode, we iterate through `topSelectNodes` from the latest visited to the oldest visited,
```

Comment:
could you add a note/todo here that this implementation is subject to change and also tag the corresponding coral issue?

Reply:
Added a note, will open an issue after this PR is merged, so I didn't tag the issue.